core/src/client/async_client/mod.rs

```diff
@@ -403,163 +620,46 @@ async fn background_task<S, R>(
-				tracing::trace!("[backend]: client prepares to send request={:?}", request);
-				match sender.send(request.raw).await {
-					Ok(_) => manager
-						.insert_pending_call(request.id, request.send_back)
-						.expect("ID unused checked above; qed"),
-					Err(e) => {
-						tracing::warn!("[backend]: client request failed: {:?}", e);
-						let _ = request.send_back.map(|s| s.send(Err(Error::Transport(e.into()))));
-					}
-				}
-			}
+		let mut submit_ping = if let Some(duration) = ping_interval {
+			Delay::new(duration).fuse()
+		} else {
+			Fuse::<Delay>::terminated()
+		};
 
-			// User called `subscribe` on the front-end.
-			Either::Left((Some(FrontToBack::Subscribe(sub)), _)) => match sender.send(sub.raw).await {
-				Ok(_) => manager
-					.insert_pending_subscription(
-						sub.subscribe_id,
-						sub.unsubscribe_id,
-						sub.send_back,
-						sub.unsubscribe_method,
-					)
-					.expect("Request ID unused checked above; qed"),
-				Err(e) => {
-					tracing::warn!("[backend]: client subscription failed: {:?}", e);
-					let _ = sub.send_back.send(Err(Error::Transport(e.into())));
+		select_biased! {
+			frontend_value = next_frontend => {
+				if let Err(err) = handle_frontend_messages(frontend_value, &mut manager, &mut sender, max_notifs_per_subscription).await {
+					tracing::warn!("{:?}", err);
+					let _ = front_error.send(err);
+					break;
 				}
 			},
-			// User dropped a subscription.
-			Either::Left((Some(FrontToBack::SubscriptionClosed(sub_id)), _)) => {
-				tracing::trace!("Closing subscription: {:?}", sub_id);
-				// NOTE: The subscription may have been closed earlier if
-				// the channel was full or disconnected.
-				if let Some(unsub) = manager
-					.get_request_id_by_subscription_id(&sub_id)
-					.and_then(|req_id| build_unsubscribe_message(&mut manager, req_id, sub_id))
-				{
-					stop_subscription(&mut sender, &mut manager, unsub).await;
+			backend_value = next_backend => {
+				if let Err(err) = handle_backend_messages::<S, R>(
+					backend_value, &mut ping_submitted, &mut manager, &mut sender, max_notifs_per_subscription
+				).await {
+					tracing::warn!("{:?}", err);
+					let _ = front_error.send(err);
+					break;
 				}
-			}
-
-			// User called `register_notification` on the front-end.
-			Either::Left((Some(FrontToBack::RegisterNotification(reg)), _)) => {
-				tracing::trace!("[backend] registering notification handler: {:?}", reg.method);
-				let (subscribe_tx, subscribe_rx) = mpsc::channel(max_notifs_per_subscription);
-
-				if manager.insert_notification_handler(&reg.method, subscribe_tx).is_ok() {
-					let _ = reg.send_back.send(Ok((subscribe_rx, reg.method)));
-				} else {
-					let _ = reg.send_back.send(Err(Error::MethodAlreadyRegistered(reg.method)));
+			},
+			_ = submit_ping => {
+				// Ping was already submitted.
+				// No activity from frontend, backend (replies or pong) for a duration of `ping_interval`.
+				if ping_submitted {
```

Comment:
I can't work out why `ping_submitted` is needed?

as far as I can see, if a message is sent or received, the select will resolve and we'll run the function again and start a brand new delay etc, so we'll already wait for the ping interval after any other event happens?

Reply:
Ooh, so you set `ping_submitted` to true and then expect it to be set back to false with a pong message before we would want to do another ping; got it!

So if the ping interval is really short, we'll also expect pongs back quicker. It may be worth adding in the docs for setting a ping interval that the client will expect a pong to be returned within that interval else it'll emit an error.